continunet/network/unet.py

```diff
@@ -0,0 +1,135 @@
+        """Convolutional block for UNet."""
+        convolutional_layer = Conv2D(
+            filters=filters,
+            kernel_size=(kernel_size, kernel_size),
+            kernel_initializer="he_normal",
+            padding="same",
+        )
+        batch_normalisation_layer = BatchNormalization()
+        relu_layer = Activation("relu")
+
+        if self.batch_normalisation:
+            return relu_layer(batch_normalisation_layer(convolutional_layer(input_tensor)))
+        return relu_layer(convolutional_layer(input_tensor))
+
+    def encoding_block(self, input_tensor, filters, kernel_size=3):
+        """Encoding block for UNet."""
+        convolutional_block = self.convolutional_block(input_tensor, filters, kernel_size)
+        max_pooling_layer = MaxPooling2D((2, 2), padding="same")
+        dropout_layer = Dropout(self.dropout)
+
+        return convolutional_block, dropout_layer(max_pooling_layer(convolutional_block))
+
+    def decoding_block(self, input_tensor, concat_tensor, filters, kernel_size=3):
+        """Decoding block for UNet."""
+        transpose_convolutional_layer = Conv2DTranspose(
+            filters, (3, 3), strides=(2, 2), padding="same"
+        )
+        skip_connection = Concatenate()(
+            [transpose_convolutional_layer(input_tensor), concat_tensor]
+        )
+        dropout_layer = Dropout(self.dropout)
+        return self.convolutional_block(dropout_layer(skip_connection), filters, kernel_size)
+
+    def build_model(self):
+        """Build the UNet model."""
+        input_image = Input(self.input_shape, name="img")
+
+        # Encoder Path
+        first_convolutional_tensor, first_feature_map = self.encoding_block(
+            input_image, self.filters * 1
+        )
+        second_convolutional_tensor, second_feature_map = self.encoding_block(
+            first_feature_map, self.filters * 2
+        )
+        third_convolutional_tensor, third_feature_map = self.encoding_block(
+            second_feature_map, self.filters * 4
+        )
+        fourth_convolutional_tensor, fourth_feature_map = self.encoding_block(
+            third_feature_map, self.filters * 8
+        )
+        latent_convolutional_tensor = self.convolutional_block(
+            fourth_feature_map, filters=self.filters * 16
+        )
+
+        # Decoder Path
+        sixth_convolutional_tensor = self.decoding_block(
+            latent_convolutional_tensor, fourth_convolutional_tensor, self.filters * 8
+        )
+        seventh_convolutional_tensor = self.decoding_block(
+            sixth_convolutional_tensor, third_convolutional_tensor, self.filters * 4
+        )
+        eighth_convolutional_tensor = self.decoding_block(
+            seventh_convolutional_tensor, second_convolutional_tensor, self.filters * 2
+        )
+        ninth_convolutional_tensor = self.decoding_block(
+            eighth_convolutional_tensor, first_convolutional_tensor, self.filters * 1
+        )
```

Comment:
You could do this as a for loop, something like
```py
current = input_image
mult = 1

# encoding 
for _ in range(3):
    filters = self.filters * mult
    mult = mult * 2
    current = self.encoding_block(current, filters) 

# convolutional layer
current = self.convolutional_block(current, self.filters * mult)

# decoder 
while mult != 1:
    mult = mult // 2
    current = self.decoding_block(current, self.filters * mult) 
```
Less room for mistakes, easier to grow.

Reply:
Have refactored this :)